usr/src/uts/armv8/os/gicv2.c

```diff
@@ -21,152 +21,291 @@
- * edge-triggered interrupts.  This is set when `gicv2_config_irq` and never
- * altered (which makes sense).
+ * In the standard configuration we don't use sub-priority at all.
  */
-static uint32_t intr_cfg[1024 / 32];
+#define	STANDARD_BPR			0x00000000
 
 /*
- * Protect access to global GIC state.
- * In the current implementation, the distributor.
+ * Required BPR is 3 bits
  */
-static lock_t gicv2_lock;
-
-/* Base addresses of the distributor */
-static volatile struct gic_dist *gic_dist;
+static uint32_t bodged_priorities[] = {
+	/*  0 */ 240,	/* Real */
+	/*  1 */ 228,	/* Fake */
+	/*  2 */ 226,	/* Fake */
+	/*  3 */ 225,	/* Fake */
+	/*  4 */ 224,	/* Real, Disk */
+	/*  5 */ 209,	/* Fake */
+	/*  6 */ 208,	/* Real */
+	/*  7 */ 192,	/* Real, NIC */
+	/*  8 */ 177,	/* Fake */
+	/*  9 */ 176,	/* Real */
+	/* 10 */ 160,	/* Real, Clock  */
+	/* 11 */ 144,	/* Real, Dispatcher */
+	/* 12 */ 132,	/* Fake */
+	/* 13 */ 130,	/* Fake */
+	/* 14 */ 129,	/* Fake */
+	/* 15 */ 128,	/* Real */
+};
 
+#define	BODGED_PRIORITY_PMR_MASK	0x000000F0
 /*
- * Mapping from cpuid to GIC target identifier
- * XXXARM: Another place we cap at 8 CPUs
+ * Configure the priority fields with 5 bits of group priority and 3 bits of
+ * subpriority. This may not actually work, as the minimums may be lower,
+ * but we do the best that we can.
+ *
+ * Even of the minimum is lower our priority mask remains valid, it's just
+ * ordering that might be affected.
  */
-static uint8_t gicv2_target[8];
+#define	BODGED_BPR			0x00000002
 
 /*
- * CPUs for which we have initialized the GIC.  Used to limit IPIs to only
- * those CPUs we can target.
+ * IPL - >GIC Priority table to use.
  */
-static cpuset_t gicv2_cpuset;
-
+static uint32_t *gicv2_prio_map;
 /*
- * Enable IRQ in the distributor, which will now be forwarded to a cpu.
+ * Mask to apply to the GIC priority when setting the priority mask register
+ * on the GIC CPU Interface.
  *
- * It is IMPLEMENTATION DEFINED whether SGIs can be enabled or disabled here,
- * we never try.
+ * Do not apply this mask when setting interrupt configuration.
  */
-static void
-gicv2_enable_irq(int irq)
+static uint32_t gicv2_prio_pmr_mask;
+
+#undef GIC_IPL_TO_PRIO
+#define	GIC_IPL_TO_PRIO(v)		(gicv2_prio_map[((v) & 0xF)])
+
+#define	GICV2_GICD_LOCK_INIT_HELD	uint64_t old_daif = read_daif(); \
+					set_daif(DAIF_SETCLEAR_IRQ); \
+					LOCK_INIT_HELD(&conf.gc_lock)
+#define	GICV2_GICD_LOCK			uint64_t old_daif = read_daif(); \
+					set_daif(DAIF_SETCLEAR_IRQ); \
+					lock_set(&conf.gc_lock)
+#define	GICV2_GICD_UNLOCK		lock_clear(&conf.gc_lock); \
+					write_daif(old_daif)
+#define	GICV2_ASSERT_GICD_LOCK_HELD	ASSERT(LOCK_HELD(&conf.gc_lock))
+
+static inline uint32_t
+gicc_read(gicv2_conf *gic, uint32_t reg)
 {
-	if (irq >= GIC_INTID_MIN_PPI) {
-		gic_dist->isenabler[GICD_IRQ_TO_ISENABLER(irq)] =
-		    GICD_IRQ_TO_ISENABLER_FIELD(irq);
-	}
+	return (i_ddi_get32(NULL, (uint32_t *)(gic->gc_gicc + reg)));
```

Comment:
You're meant to call `ddi_getXX` not the implementations `i_ddi_getXX`, DDI-wise.  Did that not work out?  (this applies throughout here).

Reply:
Aah, yes.

The reason I'm calling the implementation directly is that I have to pass a NULL `hdlp` argument. Looking at the aarch64 implementation this doesn't get checked (or used), so I'll switch over to the DDI version.

Reply:
On re-examination, I can't change this until we get startup etc. in a good enough shape to properly hook up a `ddi_acc_impl_t`. Until then we'd crash here: https://github.com/richlowe/illumos-gate/blob/arm64-gate/usr/src/uts/aarch64/os/ddi_aarch64.c#L327